### PR TITLE
PROD-515: Mystery box hi-fi claim flow

### PR DESCRIPTION
**Pull request by samott**
PROD-515: Mystery box hi-fi claim flow.

### CODE REVIEW

File: app/actions/mysteryBox.ts [hunk 1]
```diff
@@ -203,3 +271,22 @@ export async function handleSendBonk(rewardAmount: number, address: string) {
 
   return null;
 }
+
+export async function calculateTotalPrizeTokens(
```
Comment (marvinmarnold) [inline app/actions/mysteryBox.ts:275]: Can you review this ticket and update these actions: https://linear.app/gator/issue/PROD-212/lockdown-server-actions

Top of mind wishlist:
- new actions each get their own file and own tests
- helper methods not invoked directly by frontend are moved outside `/app`
- all actions have an auth mechanism

File: app/components/MysteryBox/MysteryBoxAmount.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import { numberToCurrencyFormatter } from "@/app/utils/currency";
```
Comment (marvinmarnold) [inline app/components/MysteryBox/MysteryBoxAmount.tsx:1]: Ideally all new Components would go outside app/ directory. app/ should only have layouts, pages, and routes.

File: app/components/MysteryBox/MysteryBoxPrize.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  amount: number;
+};
+
+function MysteryBoxPrize({ type, amount }: MysteryBoxPrizeProps) {
+  return (
+    <div
+      className={cn(
+        "w-[117px] h-[142px] rounded-xl p-1 flex flex-col items-center justify-around",
+        {
+          "bg-[#89E0CD]": type == "credits",
```
Comment (marvinmarnold) [inline app/components/MysteryBox/MysteryBoxPrize.tsx:21]: Consider adding these colors to tailwind.config.ts if we will reuse.

**Review by marvinmarnold — COMMENTED** [review]
Will test functionality next.

Generally, we should move as much code out of `/app/` dir as possible.

**Review by yash431garg — COMMENTED** [review]
Code looks fine to me after including @marvinmarnold reviews. But Mystery Box doesn't opens in the preview link. I have checked and mysterybox isn't created in the staging db as well.

**Comment by samott** [comment]
> Code looks fine to me after including @marvinmarnold reviews. But Mystery box doesn't opens in the preview link. I have checked and mysterybox isn't created in the staging db as well. It is kind of a known issue I have seen this issue in our previous mystery PR as well. Let me know if you need help to debug @samott.

@yash431garg I needed to fix up the staging database. I've redeployed. Please check again.

**Comment by yash431garg** [comment]
> > Code looks fine to me after including @marvinmarnold reviews. But Mystery box doesn't opens in the preview link. I have checked and mysterybox isn't created in the staging db as well. It is kind of a known issue I have seen this issue in our previous mystery PR as well. Let me know if you need help to debug @samott.
> 
> @yash431garg I needed to fix up the staging database. I've redeployed. Please check again.

Yes, It shows for me but I can't see open button. Also, what fix was needed to make it work in staging?

**Comment by yash431garg** [comment]
> > > > Code looks fine to me after including @marvinmarnold reviews. But Mystery box doesn't opens in the preview link. I have checked and mysterybox isn't created in the staging db as well. It is kind of a known issue I have seen this issue in our previous mystery PR as well. Let me know if you need help to debug @samott.
> > > 
> > > 
> > > @yash431garg I needed to fix up the staging database. I've redeployed. Please check again.
> > 
> > 
> > Yes, It shows for me but I can't see open button. Also, what fix was needed to make it work in staging?
> 
> Okay - will take a look again once I've done some refactoring. I removed the MysteryBox tables/enums so the migration could recreate them:
> 
> ```
> DELETE FROM _prisma_migrations WHERE "migration_name" = '20241126064715_new_mystery_box_schema';
> 
> DROP TABLE "MysteryBoxPrize";
> DROP TABLE "MysteryBoxTrigger";
> DROP TABLE "MysteryBox";
> DROP TYPE "EMysteryBoxStatus";
> DROP TYPE "EBoxTriggerType";
> DROP TYPE "EBoxPrizeStatus";
> DROP TYPE "EPrizeSize";
> DROP TYPE "EBoxPrizeType";
> ```

Thanks for the explanation! Also, I see this right now for your reference.
Screenshot 2024-11-30 at 4 01 25 PM

**Comment by samott** [comment]
> > > > > Code looks fine to me after including @marvinmarnold reviews. But Mystery box doesn't opens in the preview link. I have checked and mysterybox isn't created in the staging db as well. It is kind of a known issue I have seen this issue in our previous mystery PR as well. Let me know if you need help to debug @samott.
> > > > 
> > > > 
> > > > @yash431garg I needed to fix up the staging database. I've redeployed. Please check again.
> > > 
> > > 
> > > Yes, It shows for me but I can't see open button. Also, what fix was needed to make it work in staging?
> > 
> > 
> > Okay - will take a look again once I've done some refactoring. I removed the MysteryBox tables/enums so the migration could recreate them:
> > ```
> > DELETE FROM _prisma_migrations WHERE "migration_name" = '20241126064715_new_mystery_box_schema';
> > 
> > DROP TABLE "MysteryBoxPrize";
> > DROP TABLE "MysteryBoxTrigger";
> > DROP TABLE "MysteryBox";
> > DROP TYPE "EMysteryBoxStatus";
> > DROP TYPE "EBoxTriggerType";
> > DROP TYPE "EBoxPrizeStatus";
> > DROP TYPE "EPrizeSize";
> > DROP TYPE "EBoxPrizeType";
> > ```
> 
> Thanks for the explanation! Also, I see this right now for your reference. Screenshot 2024-11-30 at 4 01 25 PM

Ah, I guess you need to scroll down for the button.

**Comment by yash431garg** [comment]
> > > > > > Code looks fine to me after including @marvinmarnold reviews. But Mystery box doesn't opens in the preview link. I have checked and mysterybox isn't created in the staging db as well. It is kind of a known issue I have seen this issue in our previous mystery PR as well. Let me know if you need help to debug @samott.
> > > > > 
> > > > > 
> > > > > @yash431garg I needed to fix up the staging database. I've redeployed. Please check again.
> > > > 
> > > > 
> > > > Yes, It shows for me but I can't see open button. Also, what fix was needed to make it work in staging?
> > > 
> > > 
> > > Okay - will take a look again once I've done some refactoring. I removed the MysteryBox tables/enums so the migration could recreate them:
> > > ```
> > > DELETE FROM _prisma_migrations WHERE "migration_name" = '20241126064715_new_mystery_box_schema';
> > > 
> > > DROP TABLE "MysteryBoxPrize";
> > > DROP TABLE "MysteryBoxTrigger";
> > > DROP TABLE "MysteryBox";
> > > DROP TYPE "EMysteryBoxStatus";
> > > DROP TYPE "EBoxTriggerType";
> > > DROP TYPE "EBoxPrizeStatus";
> > > DROP TYPE "EPrizeSize";
> > > DROP TYPE "EBoxPrizeType";
> > > ```
> > 
> > 
> > Thanks for the explanation! Also, I see this right now for your reference. Screenshot 2024-11-30 at 4 01 25 PM
> 
> Ah, I guess you need to scroll down for the button.

I have zoom out 
Screenshot 2024-11-30 at 4 23 36 PM

**Comment by samott** [comment]
@as

> > > > > > > Code looks fine to me after including @marvinmarnold reviews. But Mystery box doesn't opens in the preview link. I have checked and mysterybox isn't created in the staging db as well. It is kind of a known issue I have seen this issue in our previous mystery PR as well. Let me know if you need help to debug @samott.
> > > > > > 
> > > > > > 
> > > > > > @yash431garg I needed to fix up the staging database. I've redeployed. Please check again.
> > > > > 
> > > > > 
> > > > > Yes, It shows for me but I can't see open button. Also, what fix was needed to make it work in staging?
> > > > 
> > > > 
> > > > Okay - will take a look again once I've done some refactoring. I removed the MysteryBox tables/enums so the migration could recreate them:
> > > > ```
> > > > DELETE FROM _prisma_migrations WHERE "migration_name" = '20241126064715_new_mystery_box_schema';
> > > > 
> > > > DROP TABLE "MysteryBoxPrize";
> > > > DROP TABLE "MysteryBoxTrigger";
> > > > DROP TABLE "MysteryBox";
> > > > DROP TYPE "EMysteryBoxStatus";
> > > > DROP TYPE "EBoxTriggerType";
> > > > DROP TYPE "EBoxPrizeStatus";
> > > > DROP TYPE "EPrizeSize";
> > > > DROP TYPE "EBoxPrizeType";
> > > > ```
> > > 
> > > 
> > > Thanks for the explanation! Also, I see this right now for your reference. Screenshot 2024-11-30 at 4 01 25 PM
> > 
> > 
> > Ah, I guess you need to scroll down for the button.
> 
> I have zoom out Screenshot 2024-11-30 at 4 23 36 PM

@yash431garg I made some tweaks to the layout. Please take a look when you get a chance.

File: app/actions/mysteryBox.ts (hunk 1 above)
Comment (samott) [reply app/actions/mysteryBox.ts:279]: I've separated out the actions and utilities.

File: app/actions/mysteryBox/open.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      },
+    });
+
+    if (!userWallet) {
+      release();
+      return null;
+    }
+
+    const sendTx = await sendBonkFromTreasury(
+      Number(reward?.MysteryBoxPrize[0]?.amount),
```
[the remark this answers is not in the document]
Comment (samott) [reply app/actions/mysteryBox/open.ts:66]: Indeed, I'll update it (this was existing code).

File: app/actions/mysteryBox/dismiss.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  const payload = await getJwtPayload();
+
+  if (!payload) {
+    return null;
+  }
+
+  try {
+    await prisma.$transaction(async (tx) => {
+      const updated = await tx.mysteryBox.update({
+        where: {
```
[the remark this answers is not in the document]
Comment (samott) [reply app/actions/mysteryBox/dismiss.ts:20]: Make sense. I do think actions should in general be restricted to the authenticated user.

**Comment by marvinmarnold** [comment]
It doesn't show how much BONK was won and seems like it's incorrectly processing as 0.

Expected:
Screenshot 2024-12-01 at 1 49 02 PM

Actual: 
Screenshot 2024-12-01 at 1 49 22 PM

**Comment by samott** [comment]
> It doesn't show how much BONK was won and seems like it's incorrectly processing as 0.
> 
> Expected: Screenshot 2024-12-01 at 1 49 02 PM
> 
> Actual: Screenshot 2024-12-01 at 1 49 22 PM

Hmm. Did the mechanism engine return not give a reward, perhaps?

**Comment by marvinmarnold** [comment]
> Hmm. Did the mechanism engine return not give a reward, perhaps?

Idk. But even in that case I would expect the icon of gold to be visible and say 0 BONK.

**Comment by samott** [comment]
> > Hmm. Did the mechanism engine return not give a reward, perhaps?
> 
> Idk. But even in that case I would expect the icon of gold to be visible and say 0 BONK. Screenshot 2024-12-01 at 2 40 33 PM

Shouldn't we just not show the mystery box screen if there are no prizes? It's seems kind bad to say "congrats, you won 0". The bonk badge is hidden because there could be only credits.

Also, I think we need a unique constraint on prizes so there is only one prize per token per box. Unless we need to allow that?

**Comment by marvinmarnold** [comment]
> Shouldn't we just not show the mystery box screen if there are no prizes? It's seems kinda bad to say "congrats, you won 0". The bonk badge is hidden because there could be only credits.
> Although... maybe users know to expect an empty box from time to time...
Good question. I asked the group in Linea.

> Also (unrelated), I think we need a unique constraint on prizes so there is only one prize per token per box. Unless we need to allow that?
Yea, that's a good idea. That's a fair assumption for now.

**Comment by yash431garg** [comment]
Screenshot 2024-12-02 at 11 48 15 AM
I have added few comments on the preview regarding design.

cc: @samott

**Comment by yash431garg** [comment]
https://github.com/user-attachments/assets/2c947dc5-9faf-4a21-bf90-548ba8558013
Behind screen is scrollable when mystery box screen is open.

**Comment by samott** [comment]
@yash431garg do you have a screenshot of the distorted image please?

**Comment by yash431garg** [comment]
> @yash431garg do you have a screenshot of the distorted image please?

Screenshot 2024-12-02 at 6 10 39 PM
It looks stretched and now background also looks different than figma.

**Comment by samott** [comment]
> > @yash431garg do you have a screenshot of the distorted image please?
> 
> Screenshot 2024-12-02 at 6 10 39 PM It looks stretched and now background also looks different than figma.

Fixed now - thanks.

**Comment by marvinmarnold** [comment]
Screenshot 2024-12-02 at 1 04 03 PM

Was it a change in this PR that is causing `from(" ") 1 cards!` to show?

**Comment by samott** [comment]
> Was it a change in this PR that is causing `from(" ") 1 cards!` to show?

No, but I just fixed it.

**Comment by marvinmarnold** [comment]
You probably already saw, but this is a blocker: https://linear.app/gator/issue/PROD-515/hi-fi-designs-for-receive-and-open-flow-after-claim-all-share-for-dev#comment-5540e772

**Pull request merged**